### PR TITLE
[AC-1693] Send InvoiceUpcoming Notification to Client Owners

### PR DESCRIPTION
**Pull request by amorask-bitwarden**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This PR updates our handling of the `invoice.upcoming` Stripe webhook event so that, if the incoming invoice is for an Organization, we send an email notification not only to the organization's billing address, but also to the organization's client owners.

❌ This PR is based off an open PR here: https://github.com/bitwarden/server/pull/3305. Once this PR is merged, I'll rebase this PR off `master`.

For review purposes, everything including and after the [Add Organization_ReadOwnerEmailAddresses SPROC](https://github.com/bitwarden/server/pull/3319/commits/7073814103eaf9fd90fd978f8e156eb2667b5e53) commit should be reviewed as part of this PR.

## Code changes

* [2023-10-03_00_OrganizationReadOwnerEmailAddresses.sql](https://github.com/bitwarden/server/blob/c972048f7319356e9e24ab90012e2767a024efd7/util/Migrator/DbScripts/2023-10-03_00_OrganizationReadOwnerEmailAddresses.sql): Adds a stored procedure to retrieve the email addresses of an organization's client owners.

* [IOrganizationRepository.cs](https://github.com/bitwarden/server/blob/81ad17cf75076d054de863c33165610883f179ee/src/Core/Repositories/IOrganizationRepository.cs): Adds `GetOwnerEmailAddressesById` method.

* [OrganizationRepository.cs (Dapper)](https://github.com/bitwarden/server/blob/81ad17cf75076d054de863c33165610883f179ee/src/Infrastructure.Dapper/Repositories/OrganizationRepository.cs): Adds `GetOwnerEmailAddressesById` method that invokes Stored Procedure.

* [OrganizationRepository.cs (EF)](https://github.com/bitwarden/server/blob/81ad17cf75076d054de863c33165610883f179ee/src/Infrastructure.EntityFramework/Repositories/OrganizationRepository.cs): Adds `GetOwnerEmailAddressesById` method that uses EF.

* [IMailService.cs](https://github.com/bitwarden/server/blob/fdb5c4f70a45f221d3f69aab0a5ac9aab7e81800/src/Core/Services/IMailService.cs): Adds `SendInvoiceUpcoming` overload.

* [HandlebarsMailService.cs](https://github.com/bitwarden/server/blob/fdb5c4f70a45f221d3f69aab0a5ac9aab7e81800/src/Core/Services/Implementations/HandlebarsMailService.cs): Adds `SendInvoiceUpcoming` overload.

* [NoopMailService.cs](https://github.com/bitwarden/server/blob/fdb5c4f70a45f221d3f69aab0a5ac9aab7e81800/src/Core/Services/NoopImplementations/NoopMailService.cs): Adds `SendInvoiceUpcoming` overload.

* [StripeController.cs](https://github.com/bitwarden/server/blob/c972048f7319356e9e24ab90012e2767a024efd7/src/Billing/Controllers/StripeController.cs): Updates `UpcomingInvoice` handling to send the InvoiceUpcoming email notification to both an organization's billing email and client owners separately.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Review by cturnbull-bitwarden — COMMENTED** [review]
You can update this PR to target `billing/PM-212/billing-email-syncing` instead of `master` so as to not pollute the diff with the changes in https://github.com/bitwarden/server/pull/3319. Once that PR gets merged and the branch deleted, then this PR will automatically get updated to target `master` again

File: src/Core/Tools/Enums/ReferenceEventType.cs
Comment (djsmith85) [inline src/Core/Tools/Enums/ReferenceEventType.cs:46]: I don't see this used anywhere, should this be part of some other work? I also see it on #3305 https://github.com/bitwarden/server/pull/3305/files#diff-9b469d6a985fabbb18be0d8cc97a59addf356e1174cb0510c334c0657eb88911R45-R46

**Comment by amorask-bitwarden** [comment]
> You can update this PR to target `billing/PM-212/billing-email-syncing` instead of `master` so as to not pollute the diff with the changes in #3319. Once that PR gets merged and the branch deleted, then this PR will automatically get updated to target `master` again

@cturnbull-bitwarden Good call! Done.

File: src/Infrastructure.EntityFramework/Repositories/OrganizationRepository.cs
Comment (cyprain-okeke) [inline src/Infrastructure.EntityFramework/Repositories/OrganizationRepository.cs:233]: Do we need to have .Organizations in the query since OrganizationUsers already have OrganizationId. And we might to check the query because it may return duplicate emails.

See suggested query below

```
var query = (from u in dbContext.Users
                      join ou in dbContext.OrganizationUsers
                      on u.id equals ou.UserId
                      where ou.Type == 0 && ou.Status == 2
                      group u by u.email into grouped
                      select grouped.Key);
```

File: util/Migrator/DbScripts/2023-10-03_00_OrganizationReadOwnerEmailAddresses.sql
Comment (cyprain-okeke) [inline util/Migrator/DbScripts/2023-10-03_00_OrganizationReadOwnerEmailAddresses.sql:6]: Also consider the same here

File: src/Core/Tools/Enums/ReferenceEventType.cs
Comment (amorask-bitwarden) [reply src/Core/Tools/Enums/ReferenceEventType.cs:46]: Hi @djsmith85, yeah I just added this in anticipation that it might be used for event logging somewhere down the line, but wasn't sure how. I can remove it if that's not the case.

File: src/Infrastructure.EntityFramework/Repositories/OrganizationRepository.cs
Comment (amorask-bitwarden) [reply src/Infrastructure.EntityFramework/Repositories/OrganizationRepository.cs:233]: Hey @cyprain-okeke, I've updated the LINQ to SQL in this commit based on your feedback: https://github.com/bitwarden/server/pull/3319/commits/f8354d15b59c6b773ab1848fccb5ecbf7d00755f

I also noticed that I wasn't using the `organizationId` in the WHERE clause, so I added that in.

File: util/Migrator/DbScripts/2023-10-03_00_OrganizationReadOwnerEmailAddresses.sql
Comment (amorask-bitwarden) [reply util/Migrator/DbScripts/2023-10-03_00_OrganizationReadOwnerEmailAddresses.sql:7]: @cyprain-okeke Updates are in this commit: https://github.com/bitwarden/server/commit/f8354d15b59c6b773ab1848fccb5ecbf7d00755f

File: src/Core/Tools/Enums/ReferenceEventType.cs
Comment (djsmith85) [reply src/Core/Tools/Enums/ReferenceEventType.cs]: An example on how to use it can be found here: https://github.com/bitwarden/server/blob/master/src/Admin/Controllers/OrganizationsController.cs#L227-L231
but if it's not needed/required please remove it, which will also remove the need for a review from @bitwarden/team-tools-dev 

As I see you changed the base-branch, you will need to remove it there

Comment (amorask-bitwarden) [reply src/Core/Tools/Enums/ReferenceEventType.cs]: Hi @djsmith85, apologies I might be a bit confused. I am using it in that manner in the base branch: https://github.com/bitwarden/server/blob/cc5b4ddcd4e62db344af784894976f4cba1907ee/src/Billing/Controllers/StripeController.cs#L524-L525

What I meant in my earlier comment was I didn't know if I had to wire up any additional consumption logic for the raising of the event.

Comment (djsmith85) [reply src/Core/Tools/Enums/ReferenceEventType.cs]: I have to apologize as I could have sworn I searched for the usage and only commented because I didn't find anything passing that newly created reference-event-type.

Nothing for me to review here anymore. Will have a look at #3305

**Review by cyprain-okeke — DISMISSED** [review]
LGTM

**Review by cturnbull-bitwarden — DISMISSED** [review]

**Review by cturnbull-bitwarden — APPROVED** [review]

**Pull request merged**
